staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -271,13 +271,13 @@ func (s *store) Delete(
 	if err != nil {
 		return fmt.Errorf("unable to convert output object to pointer: %v", err)
 	}
-	return s.conditionalDelete(ctx, preparedKey, out, v, preconditions, validateDeletion, cachedExistingObject)
+	return s.conditionalDelete(ctx, preparedKey, out, v, preconditions, validateDeletion, cachedExistingObject, true)
 }
 
 func (s *store) conditionalDelete(
 	ctx context.Context, key string, out runtime.Object, v reflect.Value, preconditions *storage.Preconditions,
-	validateDeletion storage.ValidateObjectFunc, cachedExistingObject runtime.Object) error {
-	getCurrentState := s.getCurrentState(ctx, key, v, false)
+	validateDeletion storage.ValidateObjectFunc, cachedExistingObject runtime.Object, transformDecode bool) error {
+	getCurrentState := s.getCurrentState(ctx, key, v, false, transformDecode)
```

Comment:
below in the for loop it is getting tricky: what does `preconditions.Check(key, origState.obj)` or `validateDeletion(ctx, origState.obj)` mean if `origState.obj` is random (or zero) ?

Reply:
when a normal deletion flow fails due to corrupt object error (transform from the storage failed, or decode failed), the registry will catch that error, and retry the delete operation with an unsafe flag on, `transformDecode` will be false in that case.

Yes, pre-condition check and admission may fail with a zero object, the registry bypasses finalizer check, preconditio check when it retries with unsafe flag on.

We could have a separate implementation for corrupt object deletion by copy pasting from the current implementation, but that would be a burden - maintaining two copies of deletion logic. I wanted to reuse the existing deletion logic from the store implementation as much as possible, thoughts?

Reply:
Best idea I have is to make `objState.obj` nil and be explicit here in the loop for that case. Then it is obvious that neither the precondition nor the validation are run.

Reply:
okay, i am setting the `objState.obj` to `nil`, but leaving the `conditionalDelete` function as is, rather I am setting preconditions to nil from the registry (the code path for unsafe delete):
https://github.com/kubernetes/kubernetes/pull/127513/commits/e2c319a229a04250ec32b6508e00c42770887af8#diff-b1c0d7416e8f7dbb4cc193c383832cf3276b4474b16e7768c7323358bf9b595eR78-R82

if we change `conditionalDelete` to skip precondition and validation check if the object is nil it might lead to bugs if there is a developer error that sets the obj to nil and we will quietly skipping precondition check; now, it will throw an explicit error, thoughts?

Reply:
that could also work, yes. Skip them if they are nil.

Reply:
For the preconditions, nil is already a valid and handled value. For the validation func we decided to add nil to the possible argument values. This is now documented in this PR.